blog/fast-portfolio-sorts/index.qmd

```diff
@@ -0,0 +1,178 @@
+
+If we want to apply the function above to each month using only `base`, then we have to first split the `data.frame` into multiple parts and `lapply()` the function to each part. After we combined the parts again to one big data.frame, we can use `aggregate()` to compute the average excess returns. 
+
+```{r}
+sort_base <- function() {
+  df_split <- split(crsp_monthly_base, crsp_monthly_base$month)
+  df_split <- lapply(df_split, function(df) {
+    df$portfolio <- assign_portfolio(df, "mktcap_lag", n_portfolios = 10)
+    return(df)
+  })
+  df <- do.call(rbind, df_split)
+  aggregate(ret_excess ~ portfolio, data = df, FUN = function(x) mean(x, na.rm = TRUE))
+}
+sort_base()
+```
+
+This approach takes about 10 seconds per execution on my machine and is in fact more than 20-times slower than the other approaches! To create a more nuanced picture for the fast and arguably more interesting approaches, I'll drop the `base` approach going forward. 
+
+If we want to perform the same logic using `dplyr`, we can use the following approach. Note that I use `as.data.frame()` for all approaches to ensure that the output format is the same for all approaches - a necessary requirement for a meaningful benchmark (otherwise code would not be equivalent).
+
+```{r}
+sort_dplyr <- function() {
+  crsp_monthly_dplyr |> 
+    mutate(
+      portfolio = assign_portfolio(
+        pick(everything()), "mktcap_lag", n_portfolios = 10),
+      by = "month"
+    ) |> 
+    group_by(portfolio) |> 
+    summarize(ret = mean(ret_excess, na.rm = TRUE)) |> 
+    as.data.frame()
+}
+sort_dplyr()
+```
+
+The equivalent approach in `data.table` looks as follows. Note that I deliberately don't use any pipe or intermediate assignments as to avoid any performance overhead that these might introduce. 
+
+```{r}
+sort_dt <- function() {
+  as.data.frame(crsp_monthly_dt[
+    , `:=`(portfolio = assign_portfolio(.SD, "mktcap_lag", n_portfolios = 10), by = month)
+    ][, .(ret = mean(ret_excess, na.rm = TRUE)), keyby = .(portfolio)])
+}
+sort_dt()
+```
+
+Lastly, I add the `dtplyr` implementation that also takes a `data.table` as input and internally converts `dplyr` code to `data.table` syntax. Note that the final `as.data.frame()` call is used to access the results and ensure that the result format is consistent with the other approaches. 
+
+```{r}
+sort_dtplyr <- function() {
+  crsp_monthly_dtplyr |> 
+    mutate(
+      portfolio = assign_portfolio(
+        pick(everything()), "mktcap_lag", n_portfolios = 10),
+      by = "month"
+    )  |> 
+    group_by(portfolio) |> 
+    summarize(ret = mean(ret_excess, na.rm = TRUE)) |> 
+    as.data.frame()
+}
+sort_dtplyr()
+```
+
+Now that we have verified that all code chunks create the same average excess returns per portfolio, we can proceed to the performance evaluation. 
```

Comment:
maybe you add an actual verification to that claim to show it is true.

Reply:
Actually, bench::mark() does this for me, so I'll rephrase